libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp

```diff
@@ -197,6 +197,18 @@ void NavEKF2_core::readMagData()
             }
         }
 
+        // detect changes to magnetometer offset parameters and reset states
+        Vector3f nowMagOffsets = _ahrs->get_compass()->get_offsets(magSelectIndex);
+        bool changeDetected = lastMagOffsetsValid && (!is_equal(nowMagOffsets.x,lastMagOffsets.x) || !is_equal(nowMagOffsets.y,lastMagOffsets.y) || !is_equal(nowMagOffsets.z,lastMagOffsets.z));
+        if (changeDetected) {
+            // zero the learned magnetometer bias states
+            stateStruct.body_magfield.zero();
+            // clear the measurement buffer
+            storedMag.reset();
+        }
+        lastMagOffsets = nowMagOffsets;
+        lastMagOffsetsValid = true;
+
```

Comment:
If we had a paradigm for event-driven programming like that, then yes. The
way things are now, probably not...

On Tue, Mar 29, 2016 at 9:08 PM, Lucas De Marchi <notifications@github.com>
wrote:

> In libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp
> <https://github.com/ArduPilot/ardupilot/pull/3814#discussion_r57834912>:
>
> > @@ -197,6 +197,18 @@ void NavEKF2_core::readMagData()
> >              }
> >          }
> >
> > +        // detect changes to magnetometer offset parameters and reset states
> > +        Vector3f nowMagOffsets = _ahrs->get_compass()->get_offsets(magSelectIndex);
> > +        bool changeDetected = lastMagOffsetsValid && (!is_equal(nowMagOffsets.x,lastMagOffsets.x) || !is_equal(nowMagOffsets.y,lastMagOffsets.y) || !is_equal(nowMagOffsets.z,lastMagOffsets.z));
> > +        if (changeDetected) {
> > +            // zero the learned magnetometer bias states
> > +            stateStruct.body_magfield.zero();
> > +            // clear the measurement buffer
> > +            storedMag.reset();
> > +        }
> > +        lastMagOffsets = nowMagOffsets;
> > +        lastMagOffsetsValid = true;
> > +
>
> Instead of doing this every time we get the mag data, couldn't we do the
> other way? In the (rare) cases the offsets are set we let EKF reset the
> states.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/ArduPilot/ardupilot/pull/3814/files/d636a2fcd1b3d7cdc98974c163fa150f2919b0e2#r57834912>
>